src/marin/post_training/rollout_worker.py

```diff
@@ -0,0 +1,332 @@
+        self.rollout_writer = config.rollout_writer
+        self.environment = config.environment
+        self.environment_name = config.environment_name
+        self.max_input_length = config.max_input_length
+        self.max_output_length = config.max_output_length
+        self.pad_token_id = config.pad_token_id
+        self.n_prompts_per_step = config.n_prompts_per_step
+        self.n_generations = config.n_generations
+        self.temperature = config.temperature
+        self.log_freq = config.log_freq
+        self.rollout_batch_size = config.rollout_batch_size
+        self.max_rollouts = config.max_rollouts
+        self.coordinator = coordinator
+
+    def _start_inference_server(self):
+        """Start the inference server in a background thread."""
+
+        def run_server():
+            loop = asyncio.new_event_loop()
+            asyncio.set_event_loop(loop)
+            try:
+                loop.run_until_complete(self.inference_server.serve_async())
+            except Exception as e:
+                logger.error(f"Inference server error: {e}")
+
+        self._server_thread = threading.Thread(target=run_server, daemon=True)
+        self._server_thread.start()
+
+    def stop(self):
+        """Stop the inference worker loop and server."""
+        self._running = False
+        if self.inference_server:
+            self.inference_server.shutdown()
+
+    def _generate_rollout_batch(self, rng) -> tuple[list[dict], dict]:
+        """Generate a set of rollout batches from the environment."""
+        jax_distributed_barrier()
+
+        # Create Levanter inference contexts
+        policy_ctx = LevanterInferenceContext(
+            self.policy_model, self.inference_server, max_tokens=self.max_output_length
+        )
+        reference_ctx = LevanterInferenceContext(
+            self.reference_model, self.inference_server, max_tokens=self.max_output_length
+        )
+
+        rl_dataset, dataset_metrics = create_dataset_from_environment(
+            environment=self.environment,
+            policy_ctx=policy_ctx,
+            reference_ctx=reference_ctx,
+            n_examples=self.n_prompts_per_step,
+            prng_key=rng,
+            n_generations=self.n_generations,
+            max_input_length=self.max_input_length,
+            max_output_length=self.max_output_length,
+            pad_token_id=self.pad_token_id,
+            mode="train",
+            temperature=self.temperature,
+        )
+        jax_distributed_barrier()
+
+        return (
+            list(rl_dataset.iterate_batches(batch_size=self.rollout_batch_size, shuffle=True, loop=False)),
+            dataset_metrics,
+        )
+
+    def run(self):
+        """Main inference worker loop."""
+        logger.info("Starting inference worker...")
+
+        rollouts_generated = 0
+        step = 0
+        rng = jax.random.PRNGKey(0)
+
+        while self._running:
+            jax_distributed_barrier()
```

Comment:
imho this should probably have a string or something to make sure it's the same barrier but eh